### PR TITLE
Design for Dynamic Sharding  to Support Reading Original Images

### PR DESCRIPTION
**Pull request by workingloong**

### CODE REVIEW

File: docs/designs/dynamic_sharding_for_images.md [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+  // Starting and ending (non-inclusive) record number.
+  int64 start = 2;
+  int64 end = 3;
+}
+```
+
+In order to split the training data into shards, we must get the size of the
+training data firstly. For the former 2 ways, we can we can get the size by
+`len(os.listdir("images"))` and reading the CSV file. For simplicity, we also
```
Comment (skydoorkai) [inline docs/designs/dynamic_sharding_for_images.md]: `len(os.listdir("images"))` only works for the first case.

File: docs/designs/dynamic_sharding_for_images.md [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+can store the image names in a CSV file for the first way to store images.
+So, the name of the shard is the CSV file. The start and end indices are
+the line number of the CSV file. When the worker gets the shard, it can read
+the images by the lines in the CSV file.
+
+It is difficult to get the size because we don't know the format of the
```
Comment (skydoorkai) [inline docs/designs/dynamic_sharding_for_images.md]: For case 3, it is ...

File: docs/designs/dynamic_sharding_for_images.md [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+description. So, users need to indicate the size of training data. We can
+design a function in the model definition and users implement the function
+to return the size.
+
+```python
+def get_training_size():
+    with open("annotations/captions_val2014.json") as f:
+        data = json.load(f)
+    return len(data["images"])
```
Comment (skydoorkai) [inline docs/designs/dynamic_sharding_for_images.md]: For COCO format, get annotation data length is more accurate, as it is possible that not all images have annotations.

File: docs/designs/dynamic_sharding_for_images.md [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        self._batch_size = 256
+
+    def after_run(self, run_context, run_values):
+        self._local_step += 1
+        if self._local_step * self._batch_size > dynamic_sharding.record_count:
+            dynamic_sharding.report_shard_done()
+
+def get_dataset():
+    def _record_generator():
+        while True:
```
Comment (skydoorkai) [inline docs/designs/dynamic_sharding_for_images.md:215]: Nee a break for while True

File: docs/designs/dynamic_sharding_for_images.md [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+### Create Dataset Using TensorFlow
+
+```python
+import tensorflow as tf
+
+global dynamic_sharding = DynamicShardingManager()
+
+class DynamicShardingHook(tf.train.SessionRunHook):
+    def __init__(self, num_worker, num_shards=100, max_steps=None):
```
Comment (skydoorkai) [inline docs/designs/dynamic_sharding_for_images.md]: what is `num_shards` ? And why num_worker and max_step are needed?

File: docs/designs/dynamic_sharding_for_images.md [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+            dynamic_sharding.report_shard_done()
+
+def get_dataset():
+    def _record_generator():
+        while True:
+            shard = dynamic_sharding.fetch_shard()
+            records = read_records(shard.start, shard.end)
+            for record in records:
+                yield record
+    return tf.data.Dataset.from_generator(_record_generator()
```
Comment (skydoorkai) [inline docs/designs/dynamic_sharding_for_images.md:224]: Also, shard level shuffle option is needed.

File: docs/designs/dynamic_sharding_for_images.md [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+
+global dynamic_sharding = DynamicShardingManager()
+
+class ImageDataset(torch.utils.data.IterableDataset):
+
+    def __init__(self, shuffle=False):
+        self._shuffle = shuffle
+
+    def __iter__(self):
+        while True:
```
Comment (skydoorkai) [inline docs/designs/dynamic_sharding_for_images.md:259]: Also no break for while True.

File: docs/designs/dynamic_sharding_for_images.md (hunk 5 above)
Comment (workingloong) [reply docs/designs/dynamic_sharding_for_images.md]: Those two arguments are redundant and I will delete them.

File: docs/designs/dynamic_sharding_for_images.md [hunk 8]
```diff
@@ -0,0 +1,207 @@
+# Dynamic Sharding Support Reading Original Images
+
+Now, users have to convert images into RecordIO for ElasticDL. In practice,
+many users read the original images to train their models. So, we need to
+support read the original data from the storage. However, the format to
```
Comment (brightcoder01) [inline docs/designs/dynamic_sharding_for_images.md]: there are various file formats for the original data.

File: docs/designs/dynamic_sharding_for_images.md [hunk 9]
```diff
@@ -0,0 +1,207 @@
+# Dynamic Sharding Support Reading Original Images
+
+Now, users have to convert images into RecordIO for ElasticDL. In practice,
+many users read the original images to train their models. So, we need to
+support read the original data from the storage. However, there are various
```
Comment (brightcoder01) [inline docs/designs/dynamic_sharding_for_images.md]: read -> reading

File: docs/designs/dynamic_sharding_for_images.md [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+    ```txt
+    |-- images
+        |--0001.png
+        |--0002.png
+        |--0003.png
+        |--0004.png
+        |--0005.png
+    ```
+
+    In the format, users may not need labels in the image compression.
```
Comment (brightcoder01) [inline docs/designs/dynamic_sharding_for_images.md:22]: For this case, how can we get the label data for training?

Comment (workingloong) [reply docs/designs/dynamic_sharding_for_images.md:22]: Sometimes, users don't need the label such as the CNN encoder model.

File: docs/designs/dynamic_sharding_for_images.md (hunk 3 above)
Comment (workingloong) [reply docs/designs/dynamic_sharding_for_images.md]: Yes, we can use `COCO` to get the image count by
```python
from pycocotools.coco import COCO
coco = COCO("annotations/caption.json")
return len(coco.anns.keys())
```

And I have updated the code snippet.

**Comment by zuston** [comment]
Any update on it? @workingloong

**Comment by workingloong** [comment]
> Any update on it? @workingloong 

Sorry for delay! We have added examples of TensorFlow and PyTorch with MNIST dataset in the model zoo.
PyTorch example:
https://github.com/sql-machine-learning/elasticdl/blob/develop/model_zoo/mnist/mnist_pytorch.py

TensorFlow 1.x example:
https://github.com/sql-machine-learning/elasticdl/blob/develop/model_zoo/mnist/mnist_tfv1_img.py

**Review by brightcoder01 — APPROVED** [review]
